### PR TITLE
Add -Z small-data-threshold

### PR DESCRIPTION
**Pull request by paulmenage**
This flag allows specifying the threshold size above which LLVM should not consider placing small objects in a `.sdata` or `.sbss` section.

Support is indicated in the target options via the
small-data-threshold-support target option, which can indicate either an
LLVM argument or an LLVM module flag.  To avoid duplicate specifications
in a large number of targets, the default value for support is
DefaultForArch, which is translated to a concrete value according to the
target's architecture.

### CODE REVIEW

**Comment by paulmenage** [comment]
This is a more targeted version of what I was trying to achieve with #116555 which would add `-Z llvm-module-flag` to enable the same goal on RISCV.

**Comment by b-naber** [comment]
Sorry for the late reply. I think this needs an [MCP](https://rustc-dev-guide.rust-lang.org/contributing.html?highlight=mcp#major-changes) before it can be merged.

**Comment by b-naber** [comment]
This could be merged now. Can you rebase?

**Comment by paulmenage** [comment]
There's still some lack of clarity on the MCP zulip stream whether this is the right way to solve the problem, versus a target feature. I'll push the rebase but it's probably best to not merge it yet.

**Comment by paulmenage** [comment]
OK, the conclusion on the zulip chat appeared to be that a target feature wasn't the right approach, so I think this is ready.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r? compiler-errors @bors r+

**Comment by klensy** [comment]
Can this emit (at least) warn if flag used with arch that isn't supported?

**Comment by compiler-errors** [comment]
Funny because I was going to ask the same thing, but then I removed it. Given that someone else cares about the same thing, let's implement that warning then.

@bors r-

**Comment by paulmenage** [comment]
Where would be the most appropriate place to put that check?

**Comment by paulmenage** [comment]
Is there a way to issue a warning if a flag's value isn't used? Given the way LLVM supports small-data thresholds in two completely different ways (arg versus module flag) we'd be duplicating the arch checks if we put a warning at the top level. If we could check after the build that the flag had been used that would be less fragile.

**Comment by JohnCSimon** [comment]
@paulmenage 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Comment by paulmenage** [comment]
I have an updated version of this PR, but I don't see the "Reopen pull request" button (presumably because I'm not an admin on rust-lang/rust).

**Comment by bjorn3** [comment]
You can only reopen if you force push the commit you had on this branch before closong the PR. This is a quirk of github.

**Comment by paulmenage** [comment]
I haven't pushed a new version of the PR though, I just have it locally. The version on github is still the version from January. But I still don't have a "Reopen pull request" button on my view of the PR page. (I see that button on closed PRs on a repo for which I am an admin)

**Comment by compiler-errors** [comment]
You can try to push now

**Comment by paulmenage** [comment]
Thanks :-)

**Comment by paulmenage** [comment]
I'm definitely a little uncomfortable with the need to modify the numerous RiscV and MIPS targets to indicate the support for `small-data-threshold` although it's cleaner than the ad-hoc support I had in the previous version of the patch.

Is there a nice way to apply default target options for an architecture? (Currently the defaults all seem to be defined for an OS rather than for an arch.)

**Comment by paulmenage** [comment]
Another option would be to have a `DefaultForArch` value of `SmallDataThresholdSupport`, which would be the default value in a target; we'd interpret `DefaultForArch` as appropriate on riscv/MIPS architectures, and as `None` elsewhere. Then there wouldn't be a need to modify most of the target files (nor existing .json target definitions)

**Comment by paulmenage** [comment]
Oops, some unintentional reformatting crept in there, I'll fix that.

**Review by compiler-errors — APPROVED** [review]

**Comment by paulmenage** [comment]
In the last update, I tweaked it to include a default based on the architecture, to avoid having to duplicate the support in many different target files.

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (klensy) [inline compiler/rustc_codegen_llvm/src/context.rs:395]: There is a `CString` for that, no need to manually add nulls.

File: tests/assembly/small_data_threshold.rs
Comment (klensy) [inline tests/assembly/small_data_threshold.rs:65]: typo

File: tests/assembly/small_data_threshold.rs
Comment (klensy) [inline tests/assembly/small_data_threshold.rs:71]: ...

File: tests/assembly/small_data_threshold.rs
Comment (klensy) [inline tests/assembly/small_data_threshold.rs:78]: This and 2 next looks like should be `MIPS-NOT`?

File: tests/assembly/small_data_threshold.rs
Comment (klensy) [inline tests/assembly/small_data_threshold.rs:67]: `:`

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (paulmenage) [reply compiler/rustc_codegen_llvm/src/context.rs:395]: Something like `let flag = CString::new(flag).unwrap()`? Or should `SmallDataThresholdSupport::LllvmModuleFlag` contain a `StaticCow<Cstr>` rather than `StaticCow<str>`?

File: tests/assembly/small_data_threshold.rs
Comment (klensy) [inline tests/assembly/small_data_threshold.rs:80]: `:`

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (klensy) [reply compiler/rustc_codegen_llvm/src/context.rs:395]: No, changing type to `StaticCow<Cstr>` is probably no needed.

Comment (paulmenage) [reply compiler/rustc_codegen_llvm/src/context.rs]: OK, tweaked to use `let flag = SmallCStr::new(flag.as_ref());`

**Comment by paulmenage** [comment]
I fixed the typos in the tests.

**Comment by compiler-errors** [comment]
Thanks @klensy for your thorough reviews as always

**Pull request merged**
